cabal-install/Distribution/Client/ProjectBuilding.hs

```diff
@@ -0,0 +1,1257 @@
+          -- TODO: [nice to have] use a proper file monitor rather than this dir exists test
+          exists <- doesDirectoryExist srcdir
+          if exists
+            then dryRunLocalPkg pkg depsBuildStatus srcdir
+            else return (BuildStatusUnpack tarball)
+      where
+        srcdir = distUnpackedSrcDirectory (packageId pkg)
+
+    dryRunLocalPkg :: ElaboratedConfiguredPackage
+                   -> ComponentDeps [BuildStatus]
+                   -> FilePath
+                   -> IO BuildStatus
+    dryRunLocalPkg pkg depsBuildStatus srcdir = do
+        -- Go and do lots of I/O, reading caches and probing files to work out
+        -- if anything has changed
+        change <- checkPackageFileMonitorChanged
+                    packageFileMonitor pkg srcdir depsBuildStatus
+        case change of
+          -- It did change, giving us 'BuildStatusRebuild' info on why
+          Left rebuild ->
+            return (BuildStatusRebuild srcdir rebuild)
+
+          -- No changes, the package is up to date. Use the saved build results.
+          Right (mipkg, buildSuccess) ->
+            return (BuildStatusUpToDate mipkg buildSuccess)
+      where
+        packageFileMonitor =
+          newPackageFileMonitor distDirLayout (packageId pkg)
+
+
+traverseInstallPlan :: forall m ipkg srcpkg iresult ifailure b.
+                       (Monad m,
+                        HasUnitId ipkg,   PackageFixedDeps ipkg,
+                        HasUnitId srcpkg, PackageFixedDeps srcpkg)
+                    => GenericInstallPlan ipkg srcpkg iresult ifailure
+                    -> (GenericPlanPackage ipkg srcpkg iresult ifailure ->
+                        ComponentDeps [b] -> m b)
+                    -> m (Map InstalledPackageId b)
+traverseInstallPlan plan0 visit =
+    go Map.empty (InstallPlan.reverseTopologicalOrder plan0)
+  where
+    go :: Map InstalledPackageId b
+       -> [GenericPlanPackage ipkg srcpkg iresult ifailure]
+       -> m (Map InstalledPackageId b)
+    go !results [] = return results
+
+    go !results (pkg : pkgs) = do
+      -- we go in the right order so the results map has entries for all deps
+      let depresults :: ComponentDeps [b]
+          depresults = fmap (map (results Map.!)) (depends pkg)
+      result <- visit pkg depresults
+      let results' = Map.insert (installedPackageId pkg) result results
+      go results' pkgs
+
+improveInstallPlanWithUpToDatePackages :: ElaboratedInstallPlan
+                                       -> BuildStatusMap
+                                       -> ElaboratedInstallPlan
+improveInstallPlanWithUpToDatePackages installPlan pkgsBuildStatus =
+    replaceWithPreInstalled installPlan
+      [ (installedPackageId pkg, mipkg, buildSuccess)
+      | InstallPlan.Configured pkg
+          <- InstallPlan.reverseTopologicalOrder installPlan
+      , let ipkgid = installedPackageId pkg
+      , BuildStatusUpToDate mipkg buildSuccess <- [pkgsBuildStatus Map.! ipkgid]
```

Comment:
Yes ok fair enough. I'll use a `let Just blah = Map.lookup` style so we get a clear exception with a location.